smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/lang/DocumentationPreprocessor.kt

```diff
@@ -21,17 +26,226 @@ class DocumentationPreprocessor : KotlinIntegration {
+                    // models. Ensure we strip those now, the parser is VERY forgiving and will mistreat any sequences
+                    // of characters that happen to form tags as such.
+                    val sanitizedDoc = trait.value
+                        .applyWithin("<code>", "</code>", String::escapeHtml)
+                        .applyWithin("<pre>", "</pre>", String::escapeHtml)
+                    val docs = toKdoc(sanitizedDoc)
                     DocumentationTrait(docs, trait.sourceLocation)
                 }
                 else -> trait
             }
         }
     }
 
-    // KDoc comments use inline markdown. Replace square brackets with escaped equivalents so that they
-    // are not rendered as invalid links
-    private fun sanitize(str: String): String =
-        str.replace("[", "&#91;")
-            .replace("]", "&#93;")
+    private fun toKdoc(doc: String): String {
+        val parsed = parseClean(doc)
+
+        val renderer = MarkdownRenderer()
+        parsed.body().traverse(renderer)
+        return renderer.text
+    }
+
+    private fun parseClean(rawDoc: String): Document {
+        val parsed = Jsoup.parse(rawDoc)
+
+        parsed.body().stripBlankTextNodes()
+
+        return parsed
+    }
+
+    private class MarkdownRenderer : NodeVisitor {
+        companion object {
+            const val SUBLIST_INDENT = "   "
+            const val PREFORMAT_MARKER = "`"
+            const val BOLD_MARKER = "**"
+            const val ITALIC_MARKER = "*"
+        }
+
+        var text: String = ""
+            private set
+
+        private var bufferedAnchorHref: String = ""
+        private var bufferedAnchorText: String = ""
+        private var listPrefix: String = ""
+
+        override fun head(node: Node, depth: Int) {
+            if (node is TextNode) {
+                if (node.parentNode()?.nodeName() == "a") {
+                    bufferedAnchorText = node.markdownText()
+                } else {
+                    text += node.markdownText()
+                }
+                return
+            }
+
+            when (node.nodeName()) {
+                "a" -> {
+                    if (node.hasAttr(("href"))) {
+                        bufferedAnchorHref = node.attr("href")
+                    }
+                }
+                "li" -> {
+                    // If this list item holds a sublist, then we essentially just want to line break right away and
+                    // render the nested list as normal.
+                    val prefix = if (node.childNode(0).nodeName() == "ul") "\n" else ""
+                    text += "$listPrefix+ $prefix"
+                }
+                "code", "pre" -> {
+                    text += PREFORMAT_MARKER
+                }
+                "b", "strong" -> {
+                    text += BOLD_MARKER
+                }
+                "i", "em" -> {
+                    text += ITALIC_MARKER
+                }
```

Comment:
**Style**: braces and newlines are not required for single-statement cases (but feel free to keep them when a line is too long or requires an adjacent comment). Example:

```kotlin
"code", "pre" -> text += PREFORMAT_MARKER
"b", "strong" -> text += BOLD_MARKER
"i", "em" -> text += ITALIC_MARKER
```